xla/service/gpu/gpu_all_gather_optimizer.cc

```diff
@@ -0,0 +1,106 @@
+
+#include "xla/service/gpu/gpu_all_gather_optimizer.h"
+
+#include "xla/hlo/ir/hlo_casting_utils.h"
+#include "xla/hlo/ir/hlo_instruction.h"
+#include "xla/hlo/ir/hlo_instructions.h"
+#include "xla/hlo/ir/hlo_module.h"
+#include "xla/hlo/ir/hlo_opcode.h"
+#include "xla/service/collective_ops_utils.h"
+
+namespace xla {
+namespace gpu {
+
+StatusOr<bool> AllGatherOptimizer::Run(
+    HloModule* module,
+    const absl::flat_hash_set<absl::string_view>& execution_threads) {
+
+  bool changed = false;
+  for (HloComputation* computation :
+       module->MakeNonfusionComputations(execution_threads)) {
+    for (HloInstruction* instruction :
+         computation->MakeInstructionPostOrder()) {
+      if (!HloOpcodeIsBinaryCommutative(instruction->opcode())) {
+        continue;
+      }
+
+      HloInstruction* left_op = instruction->mutable_operand(0);
+      HloInstruction* right_op = instruction->mutable_operand(1);
+
+      if (right_op->opcode() != HloOpcode::kAllGather ||
+          left_op->opcode() != HloOpcode::kAllGather) {
+        VLOG(2) << "Binary op's operands are not all-gather deduced types.";
+        continue;
+      }
+
+      auto* left_all_gather = Cast<HloAllGatherInstruction>(left_op);
+      auto* right_all_gather = Cast<HloAllGatherInstruction>(right_op);
+
+      if (right_all_gather->constrain_layout() !=
+              left_all_gather->constrain_layout() ||
+          right_all_gather->use_global_device_ids() !=
+              left_all_gather->use_global_device_ids() ||
+          !ReplicaGroupsEqual(right_all_gather->replica_groups(),
+                              left_all_gather->replica_groups())) {
+        VLOG(2) << "The right and left all-gather ops are not compatible "
+                   "to merge. ";
+        continue;
+      }
+
+      if (right_all_gather->user_count() != 1 ||
+          left_all_gather->user_count() != 1) {
+        VLOG(2) << "all-gather user_count > 1 ";
+        continue;
+      }
+      auto index_in_full_shape =
+          computation->AddInstruction(HloInstruction::CreateBinary(
+              right_all_gather->operand(0)->shape(), instruction->opcode(),
+              left_all_gather->mutable_operand(0),
+              right_all_gather->mutable_operand(0)));
+
+      int64_t all_gather_dimension =
+          Cast<HloAllGatherInstruction>(right_all_gather)
+              ->all_gather_dimension();
+
+      auto combined = HloInstruction::CreateAllGather(
+          left_all_gather->shape(), {index_in_full_shape}, all_gather_dimension,
+          left_all_gather->replica_groups(),
+          /*constrain_layout=*/false, left_all_gather->channel_id(),
+          Cast<HloAllGatherInstruction>(left_all_gather)
+              ->use_global_device_ids());
+
+      TF_RETURN_IF_ERROR(computation->ReplaceWithNewInstruction(
+          instruction, std::move(combined)));
+
+      auto check_instruction = [&](HloInstruction* inst) {
```

Comment:
@kushanam I though about this again and I think we should remove this code altogether. If XLA's base machinery thinks that the old all-gathers are ok to remove, it will do so as a part of ReplaceWithNewInstruction above. And if it does not think its safe, I don't think an individual pass should override that.

@sherhut WDYT?

Reply:
Removed the node removal and fixed the tests accordingly. @jurahul and @sherhut please have a look when you got a chance.

Reply:
I agree, all reduces should disappear by themselves if they are no longer used. As we rule out multiple users for the all-gathers, I wonder when they actually would not be removed?

In any case, this seems btter.